libraries/SITL/SIM_GPS.cpp

```diff
@@ -1014,6 +1017,311 @@ uint32_t GPS::CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_t crc
+    pos_time.pos_flags_1 = pos_flags_1;
+    pos_time.pos_flags_2 = pos_flags_2;
+    pos_time.time_week_ms = htobe32(gps_tow.ms);
+    pos_time.time_week = htobe16(gps_tow.week);
+    pos_time.num_sats = d->have_lock ? _sitl->gps_numsats[instance] : 3;
+
+    constexpr uint8_t GSOF_POS_TYPE = 0x02;
+    constexpr uint8_t GSOF_POS_LEN = 0x18;
+
+    int64_t lat_i64 { 0 };
+    const auto lat_rad = d->latitude * DEG_TO_RAD_DOUBLE;
+    memcpy(&lat_i64, &lat_rad, sizeof(lat_i64));
+
+    int64_t lng_i64 { 0 };
+    const auto lng_rad = d->longitude * DEG_TO_RAD_DOUBLE;
+    memcpy(&lng_i64, &lng_rad, sizeof(lng_i64));
+
+    int64_t alt_i64 { 0 };
+    const double alt_m = static_cast<double>(d->altitude);
+    memcpy(&alt_i64, &alt_m, sizeof(alt_i64));
+
+    // TODO fix constness 
+    // Method 1: Use designated initializer (or aggregate initialization)
+    // Not working, perhaps because PACKED?
+    // error: designated initializers cannot be used with a non-aggregate type ‘SITL::GPS::update_gsof(const SITL::GPS::gps_data*)::gsof_pos
+    // Method 2: Initialize in-class as const
+    // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96184
+    // Perhaps a bug in gcc? 
+    // error: ../../libraries/SITL/SIM_GPS.cpp:1096:24: error: use of local variable with automatic storage from containing function
+    //
+    // https://godbolt.org/z/7oeMfMzj6
+    struct PACKED gsof_pos {
+        const uint8_t OUTPUT_RECORD_TYPE { GSOF_POS_TYPE }; 
+        const uint8_t RECORD_LEN { GSOF_POS_LEN };
+        uint64_t lat;
+        uint64_t lng;
+        uint64_t alt;
+    } pos {};
+    static_assert(sizeof(pos.lat) == sizeof(lat_i64), "Unexpected double format");
+    // pos.OUTPUT_RECORD_TYPE = GSOF_POS_TYPE;
+    // pos.RECORD_LEN = GSOF_POS_LEN;
+    pos.lat = htobe64(lat_i64);
+    pos.lng = htobe64(lng_i64);
+    pos.alt = htobe64(alt_i64);
+
+    static_assert(sizeof(gsof_pos) - (sizeof(gsof_pos::OUTPUT_RECORD_TYPE) + sizeof(gsof_pos::RECORD_LEN)) == GSOF_POS_LEN); 
+
+    // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Velocity.html
+    constexpr uint8_t GSOF_VEL_TYPE = 0x08;
+     // use the smaller packet by ignoring local coordinate system
+    constexpr uint8_t GSOF_VEL_LEN = 0x0D;
+
+    struct PACKED gsof_vel {
+        const uint8_t OUTPUT_RECORD_TYPE { GSOF_VEL_TYPE }; 
+        const uint8_t RECORD_LEN { GSOF_VEL_LEN };
+        // TODO
+        // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Velocity%20flags
+        uint8_t flags;
+        uint32_t horiz_m_p_s;
+        uint32_t heading_rad;
+        uint32_t vertical_m_p_s;
+    } vel {};
+    static_assert(sizeof(gsof_vel) - (sizeof(gsof_vel::OUTPUT_RECORD_TYPE) + sizeof(gsof_vel::RECORD_LEN)) == GSOF_VEL_LEN);
+
+    int32_t speed_i32 { 0 };
+    const auto speed_2d = d->speed_2d();
+    memcpy(&speed_i32, &speed_2d, sizeof(speed_2d));
```

Comment:
This looks like an inappropriate use of `auto`.

`auto` is great, but *not* when you are then taking `sizeof` to that type and copying into a fixed-length buffer!

Replace this with a similar pack function as above, anyway.

Reply:
It's now in the packing function, uniquely named and passed by reference, and protected by a static assert. It would be extremely difficult to abuse now.